syntax = "proto3";
package minecpp.proto.event.clientbound;

import "google/protobuf/any.proto";
import "minecpp/proto/common/Common.proto";
import "minecpp/proto/common/Vector.proto";
import "minecpp/proto/player/Id.proto";
import "minecpp/proto/player/Player.proto";
import "minecpp/proto/entity/Entity.proto";
import "minecpp/proto/chunk/Chunk.proto";

message Event {
  google.protobuf.Any payload = 1;
  oneof recipient {
    RecipientSinglePlayer single_player = 2;
    RecipientMultiplePlayers multiple_players = 3;
    RecipientAllPlayers all_players = 4;
    RecipientExcluding excluding = 5;
  }
}

message RecipientSinglePlayer {
  player.PlayerId player_id = 1;
}

message RecipientMultiplePlayers {
  repeated player.PlayerId player_ids = 1;
}

message RecipientAllPlayers {
}

message RecipientExcluding {
  player.PlayerId player_id = 1;
}

message AddPlayer {
  player.PlayerId player_id = 1;
  string name = 2;
  common.GameMode game_mode = 3;
  uint32 ping = 4;
}

message SpawnEntity {
  entity.Entity entity = 1;
}

message SpawnPlayer {
  player.PlayerId player_id = 1;
  entity.PlayerEntity entity = 2;
}

message EntityMove {
  uint32 entity_id = 1;
  common.Vector3i movement = 2;
  common.Rotation rotation = 3;
  bool is_on_ground = 4;
}

message EntityLook {
  player.PlayerId player_id = 1;
  uint32 entity_id = 2;
  common.Rotation rotation = 3;
  bool is_on_ground = 4;
}

message PlayerPositionRotation {
  common.Vector3 position = 1;
  common.Rotation rotation = 2;
}

message SetSpawnPosition {
  uint64 position = 1;
  float angle = 2;
}

message Chat {
  string message = 1;
  int32 type = 2;
}

message RemovePlayer {
  player.PlayerId player_id = 1;
  uint32 entity_id = 2;
}

message UpdateBlock {
  uint64 block_position = 1;
  uint32 state = 2;
}

message AnimateEntity {
  uint32 entity_id = 1;
  common.EntityAnimation animation = 2;
}

message AcknowledgeBlockChange {
  int32 sequence_id = 1;
}

message LoadTerrain {
  common.ChunkPosition central_chunk = 1;
  repeated common.ChunkPosition coords = 2;
}

message TransferPlayer {
  uint64 engine_instance_id = 2;
}

message UpdatePlayerAbilities {
  player.PlayerId player_id = 1;
  bool invulnerable = 2;
  bool is_flying = 3;
  bool allow_flying = 4;
  bool creative_mode = 5;
  float walk_speed = 7;
  float fly_speed = 8;
}

message UnloadChunk {
  player.PlayerId player_id = 1;
  common.ChunkPosition chunk_position = 2;
}

message PlayerList {
  repeated player.Status list = 1;
}

message EntityList {
  repeated entity.Entity entities = 1;
  repeated entity.PlayerEntity player_entities = 2;
}

message SetEntityVelocity {
  uint32 entity_id = 1;
  common.Vector3i velocity = 2;
}

message AcceptPlayer {
  common.Gameplay gameplay = 1;
  entity.Abilities abilities =  2;
  player.Player player = 3;
}

message DenyPlayer {
  string denial_reason = 1;
}

message SetInventorySlot {
  uint32 slot_id = 1;
  common.Slot slot = 2;
}

message SectionBlockLight {
  int32 y = 1;
  bytes block_light = 2;
}

message ChunkBlockLight {
  common.ChunkPosition position = 1;
  repeated SectionBlockLight sections = 2;
}

message UpdateBlockLight {
  repeated ChunkBlockLight block_light = 1;
}

message ChunkData {
  chunk.Chunk chunk = 1;
  bool is_initial_chunk = 2;
}

message SetCenterChunk {
  common.ChunkPosition position = 1;
}

message SetEntityEquipment {
  uint32 entity_id = 1;
  entity.EquipmentSlot equipment_slot = 2;
  common.Slot item = 3;
}

message SetHealth {
  float health = 1;
  int32 food = 2;
  float food_saturation = 3;
}

message CollectItem {
  uint32 collected_entity_id = 1;
  uint32 collector_entity_id = 2;
  uint32 count = 3;
}

message RemoveEntity {
  uint32 entity_id = 1;
}

message TeleportEntity {
  uint32 entity_id = 1;
  common.Vector3 position = 2;
  common.Rotation rotation = 3;
  bool is_on_ground = 4;
}

message DisplayDeathScreen {
  uint32 victim_entity_id = 1;
  uint32 killer_entity_id = 2;
  string death_message = 3;
};

message Respawn {
  string dimension_type = 1;
  string dimension_name = 2;
  int64 hashed_seed = 3;
  common.GameMode game_mode = 4;
  common.GameMode previous_game_mode = 5;
  bool is_debug = 6;
  bool is_flat = 7;
  bool copy_metadata = 8;
  bool has_death_location = 9;
  string death_dimension = 10;
  common.Vector3 death_position = 11;
};

message SetAbilities {
  entity.Abilities abilities = 1;
}

message RawMessage {
  bytes message_data = 1;
}
